### PR TITLE
Fuzz tests for inclusion and consistency proofs

### PR DESCRIPTION
**Pull request by hickford**
Fuzz tests for inclusion and consistency proofs

### CODE REVIEW

**Comment by mhutchinson** [comment]
What's the motivation behind removing the 1.17 tests? The go.mod file still lists this version as 1.16 compatible. I'm up for bumping this to go1.17 to be consistent with trillian (and soon, trillian-examples: https://github.com/google/trillian-examples/pull/669), but jumping to 1.18 invites using newer language features that may exclude people on older tooling. Not to say we absolutely can't do it, as we don't have an explicit policy, but we've tried to keep support for go versions in the order of a year old thus far.

**Comment by AlCutter** [comment]
> 

It's the fuzz testing support - it's new in go1.18 `testing` lib, so the test would no-longer compile with older releases.

Agreed we should chat about the trade off.

**Comment by mhutchinson** [comment]
I've just seen #32 and think that this PR with the changes to keep support for 1.17 would make this a super 👍

File: testonly/tree_fuzz_test.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/transparency-dev/merkle/proof"
+)
+
+// Compute and verify consistency proofs
+func FuzzConsistencyProof(f *testing.F) {
+	for size := 0; size <= 8; size++ {
+		for size2 := 0; size2 <= size; size2++ {
+			for size1 := 0; size1 <= size2; size1++ {
```
Comment (mhutchinson) [inline testonly/tree_fuzz_test.go:18]: I realise that you're following the naming in the files under test, but trying to understand this initialization loop order (size, size2, size1) derailed me. Perhaps `size`, `from`, `to` (or `start` and `end`) would be clearer in the tests?

File: testonly/tree_fuzz_test.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		entries := genEntries(size)
+		tree := newTree(entries)
+		got, err := tree.InclusionProof(index, size)
+		t.Logf("proof=%v", got)
+		if err != nil {
+			t.Error(err)
+		}
+		want := refInclusionProof(entries, index, tree.hasher)
+		if diff := cmp.Diff(got, want, cmpopts.EquateEmpty()); diff != "" {
+			t.Fatalf("InclusionProof(%d, %d): diff (-got +want)\n%s", index, size, diff)
```
Comment (mhutchinson) [inline testonly/tree_fuzz_test.go:107]: Error/Errorf has been used everywhere else. Any reason this one needs to be fatal?

**Review by mhutchinson — APPROVED** [review]

**Comment by mhutchinson** [comment]
@AlCutter we could do with making a call on whether the init loops added to this, or the file-based approach in #35 should be our approach to seeding fuzz tests.

**Comment by mhutchinson** [comment]
@hickford can you take a look at & respond to the open comments on this PR. Once addressed one way or another, I can merge this.

File: testonly/tree_fuzz_test.go (hunk 1 above)
Comment (hickford) [reply testonly/tree_fuzz_test.go]: Good idea, done.

**Pull request merged**